### PR TITLE
Implement basic WAMP-CRA

### PR DESCRIPTION
**Pull request by om26er**
Password salting is missing, we'll do that in another PR

### CODE REVIEW

**Comment by om26er** [comment]
CI is passing now, so this is probably good to go.

File: autobahn/src/main/java/io/crossbar/autobahn/wamp/auth/ChallengeResponseAuth.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        this.secret = secret;
+        this.authextra = authextra;
+        try {
+            SecretKeySpec secretKey = new SecretKeySpec(secret.getBytes("UTF-8"), "HmacSHA256");
+            sha256_HMAC = Mac.getInstance("HmacSHA256");
+            sha256_HMAC.init(secretKey);
+            if (Platform.isAndroid()) {
+                mBase64Class = Class.forName("android.util.Base64");
+            } else {
+                // XXX - this adds a dependency on Java8 for non-android systems.
```
Comment (oberstet) [inline autobahn/src/main/java/io/crossbar/autobahn/wamp/auth/ChallengeResponseAuth.java]: this is fine for now (relying on java8)

**Comment by oberstet** [comment]
great work! so we have anonymous, ticket and wamp-cra. cool. the only important one we still need is wamp-cryptosign (we need it for CFC/XBR app ..)

**Comment by oberstet** [comment]
wamp-cryptosign is obviously more complex .. but the first question is: what NaCl library to use?

**Comment by om26er** [comment]
> wamp-cryptosign is obviously more complex .. but the first question is: what NaCl library to use?

I have started looking into that now. What we can do initially is start with 'any' NaCl lib and once we get things working, I can probably abstract things out for easy replacement.

(My above statement might just be null given I don't (yet) know how wamp-cryptosign is supposed to work in practice.)

**Comment by oberstet** [comment]
> start with 'any' NaCl lib and once we get things working, I can probably abstract things out for easy replacement.

good approach!

here is the code that computes the signature for a WAMP-cryptosign challenge received in ABPy:

https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/cryptosign.py#L452

**Comment by oberstet** [comment]
one note on above code (the ABPy I linked):

https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/cryptosign.py#L483 is async, because the ABPy implementation is capable of talking to an SSH agent for the actual key signing. That way, the private key is never ever touched at all by the ABPy program - only by the SSH agent, which in turn might talk to a hardware based crypto thingy.

We don't need that for ABPy, so you can make computing the signature completely synchronous code. The signing comsumes some CPU cycles, but not as much as it blocks the reactor for too long (Ed25519 is super efficient algo).

**Pull request merged**